cmsl1t/config.py

```diff
@@ -1,40 +1,173 @@
+event itself. E.g. ``cmsl1t.recalc.met.l1MetNot28`` reads in
+``event.caloTowers`` and creates a new object, ``event.l1MetNot28``, that can
+then be accessed by all analyzers.
+
+.. code-block:: yaml
+
+     modifiers:
+       - cmsl1t.recalc.met.l1MetNot28:
+           in: event.caloTowers
+           out: event.l1MetNot28
+       - cmsl1t.recalc.met.l1MetNot28HF:
+           in: event.caloTowers
+           out: event.l1MetNot28HF
+
+Next, you can specify if you want progress information (e.g. a progress bar)
+and how often this information is updated (``report_every`` in units of
+events).
+
+.. code-block:: yaml
+
+     progress_bar:
+       report_every: 1000
+     # or to switch it off
+     # progress_bar:
+     #   enable: False
+
+
+And finally the output section describes where the output, usually ROOT files,
+is stored. The ```template`` entry is composed of a list of paths that are
+joined to create the full output file. The template expects the following named
+parameters:
+
+* ``date``
+* ``sample_name``
+* ``run_number``
+* ``trigger_name``
+
+which are automatically filled by the config parser
+
+
+.. code-block:: yaml
+
+   output:
+     # template is a list here that is joined (os.path.join) in the config
+     # parser
+     template:
+        - benchmark/new
+        - "{date}_{sample_name}_run-{run_number}_{trigger_name}"
+'''
 from __future__ import print_function
 import yaml
 import os
 from datetime import datetime
+import logging
+from cmsl1t.utils import module
+from copy import deepcopy
 
+logger = logging.getLogger(__name__)
 TODAY = datetime.now().timetuple()
 
 
-def validate_sections(expectedSections, configPart):
-    return sorted(expectedSections) == sorted(configPart.keys())
-
-
 def get_unique_out_dir(outdir=None, revision=1):
     full_outdir = outdir + "-rev_{rev}".format(rev=revision)
     if os.path.isdir(full_outdir):
         return get_unique_out_dir(outdir, revision + 1)
     return full_outdir
 
 
+def resolve_file_paths(paths):
+    from cmsl1t.utils.root_glob import glob
+    all_files = []
+    for p in paths:
+        all_files.extend(glob(p))
+    return all_files
+
+
 class ConfigParser(object):
     SECTIONS = ['general', 'input', 'analysis', 'output']
 
     def __init__(self):
         self.config = {}
+        self.config_errors = []
 
     def read(self, input_file):
         cfg = yaml.load(input_file)
+        self._read_config(cfg)
+
+    def _read_config(self, cfg):
         cfg['general'] = dict(version=cfg['version'], name=cfg['name'])
         del cfg['version'], cfg['name']
 
-        if not validate_sections(ConfigParser.SECTIONS, cfg):
-            # TODO: improve
-            raise IOError('Invalid config')
-
+        input_files = cfg['input']['files']
+        input_files = resolve_file_paths(input_files)
+        cfg['input']['files'] = input_files
         self.config = cfg
         self.__fill_output_template()
 
+        if not self.is_valid():
```

Comment:
I think this should probably be called earlier in this method, since we already start accessing things within the config file.  Should at least be before  `__fill_output_template()` I think.  Otherwise, I think we should improve the error handling here, so either switch from using `[key]` to get values from the config dict and move to using `get(key,default)` or put everything before the `is_valid` check in a `try ... except` block.